File: app/src/main/java/com/srijayant/spendscope/domain/ExpenseParser.java
package com.srijayant.spendscope.domain;

import com.srijayant.spendscope.model.Expense;
import com.srijayant.spendscope.model.ExpenseCategory;

import java.math.BigDecimal;
import java.time.Instant;
import java.util.Locale;
import java.util.Optional;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public final class ExpenseParser {
    private static final Pattern EXPENSE_SIGNAL = Pattern.compile(
            "\\b(debited|spent|paid|purchase(?:d)?|charged|withdrawn|withdrawal|"
                    + "payment\\s+of|sent\\s+(?:to|via)|transferred\\s+to|upi\\s+(?:txn|payment))\\b",
            Pattern.CASE_INSENSITIVE
    );
    private static final Pattern CREDIT_SIGNAL = Pattern.compile(
            "\\b(credited\\s+to|payment\\s+received|amount\\s+received|refund(?:ed)?|"
                    + "reversal|reversed|cashback\\s+(?:received|credited)|deposited)\\b",
            Pattern.CASE_INSENSITIVE
    );
    private static final Pattern AMOUNT = Pattern.compile(
            "(?:₹|rs\\.?|inr)\\s*([0-9][0-9,]*(?:\\.[0-9]{1,2})?)"
                    + "|([0-9][0-9,]*(?:\\.[0-9]{1,2})?)\\s*(?:₹|inr\\b)",
            Pattern.CASE_INSENSITIVE
    );
    private static final Pattern MERCHANT = Pattern.compile(
            "\\b(?:at|to|towards)\\s+([a-z0-9][a-z0-9 .&'@_-]{1,50}?)"
                    + "(?=\\s+(?:on|via|using|ref|reference|upi|txn|transaction|avl|"
                    + "available|bal|a/c|account|with)\\b|[().,;]|$)",
            Pattern.CASE_INSENSITIVE
    );
    private static final Pattern WHITESPACE = Pattern.compile("\\s+");

    public Optional<Expense> parse(
            long messageId,
            String sender,
            String body,
            long timestampMillis
    ) {
        if (body == null || body.isBlank() || timestampMillis <= 0) {
            return Optional.empty();
        }

        String normalized = WHITESPACE.matcher(body).replaceAll(" ").trim();
        if (!EXPENSE_SIGNAL.matcher(normalized).find() || CREDIT_SIGNAL.matcher(normalized).find()) {
            return Optional.empty();
        }

        Matcher amountMatcher = AMOUNT.matcher(normalized);
        if (!amountMatcher.find()) {
            return Optional.empty();
        }

        String amountText = amountMatcher.group(1) != null
                ? amountMatcher.group(1)
                : amountMatcher.group(2);
        BigDecimal amount;
        try {
            amount = new BigDecimal(amountText.replace(",", ""));
        } catch (NumberFormatException ignored) {
            return Optional.empty();
        }
        if (amount.compareTo(BigDecimal.ZERO) <= 0) {
            return Optional.empty();
        }

        String merchant = extractMerchant(normalized, sender);
        ExpenseCategory category = categorize(normalized + " " + merchant);
        return Optional.of(new Expense(
                messageId,
                amount,
                merchant,
                category,
                Instant.ofEpochMilli(timestampMillis)
        ));
    }

    private String extractMerchant(String body, String sender) {
        Matcher matcher = MERCHANT.matcher(body);
        if (matcher.find()) {
            String merchant = cleanMerchant(matcher.group(1));
            if (!merchant.isBlank() && !merchant.equalsIgnoreCase("your account")) {
                return merchant;
            }
        }

        if (sender != null && !sender.isBlank()) {
            String cleanedSender = sender.replaceAll("[^A-Za-z0-9 -]", "").trim();
            if (cleanedSender.matches(".*[A-Za-z].*")) {
                return cleanedSender;
            }
        }
        return "Transaction alert";
    }

    private String cleanMerchant(String merchant) {
        String cleaned = merchant
                .replaceFirst("(?i)^vpa\\s+", "")
                .replaceAll("(?i)\\s+(?:has|for|dated)$", "")
                .replaceAll("\\s+", " ")
                .trim();
        if (cleaned.length() > 36) {
            cleaned = cleaned.substring(0, 36).trim();
        }
        if (cleaned.equals(cleaned.toUpperCase(Locale.ROOT))) {
            StringBuilder title = new StringBuilder();
            for (String word : cleaned.toLowerCase(Locale.ROOT).split(" ")) {
                if (!word.isEmpty()) {
                    if (title.length() > 0) {
                        title.append(' ');
                    }
                    title.append(Character.toUpperCase(word.charAt(0))).append(word.substring(1));
                }
            }
            return title.toString();
        }
        return cleaned;
    }

    private ExpenseCategory categorize(String text) {
        String value = text.toLowerCase(Locale.ROOT);
        if (containsAny(value, "atm", "cash withdrawal", "withdrawn")) {
            return ExpenseCategory.CASH;
        }
        if (containsAny(value, "grocery", "groceries", "supermarket", "bigbasket",
                "blinkit", "zepto", "dmart", "jiomart", "more retail")) {
            return ExpenseCategory.GROCERIES;
        }
        if (containsAny(value, "swiggy", "zomato", "restaurant", "cafe", "coffee",
                "food", "bakery", "dhaba", "biryani")) {
            return ExpenseCategory.FOOD;
        }
        if (containsAny(value, "fuel", "petrol", "diesel", "iocl", "hpcl", "bpcl",
                "indian oil", "hindustan petroleum", "bharat petroleum", "fastag")) {
            return ExpenseCategory.FUEL;
        }
        if (containsAny(value, "irctc", "railway", "airline", "flight", "hotel",
                "makemytrip", "cleartrip", "goibibo", "yatra", "airbnb", "redbus")) {
            return ExpenseCategory.TRAVEL;
        }
        if (containsAny(value, "uber", "ola", "rapido", "metro", "cab", "auto fare",
                "namma yatri")) {
            return ExpenseCategory.TRANSPORT;
        }
        if (containsAny(value, "amazon", "flipkart", "myntra", "shopping", "retail",
                "store", "mall", "meesho", "ajio")) {
            return ExpenseCategory.SHOPPING;
        }
        if (containsAny(value, "house rent", "monthly rent", "rent payment")) {
            return ExpenseCategory.RENT;
        }
        if (containsAny(value, " emi ", "loan repayment", "loan payment", "home loan",
                "personal loan", "vehicle loan")) {
            return ExpenseCategory.EMI;
        }
        if (containsAny(value, "electricity", "broadband", "recharge", "utility",
                "insurance", "postpaid", "water bill", "gas bill", "mobile bill",
                "dth", "bescom", "mseb", "tata power")) {
            return ExpenseCategory.BILLS;
        }
        if (containsAny(value, "hospital", "pharmacy", "medical", "clinic", "doctor",
                "health", "apollo", "medplus")) {
            return ExpenseCategory.HEALTH;
        }
        if (containsAny(value, "netflix", "spotify", "cinema", "movie", "gaming",
                "bookmyshow", "hotstar", "prime video")) {
            return ExpenseCategory.ENTERTAINMENT;
        }
        if (containsAny(value, "school", "college", "tuition", "course", "udemy",
                "education", "exam fee", "books")) {
            return ExpenseCategory.EDUCATION;
        }
        if (containsAny(value, "upi", "imps", "neft", "transferred", "sent to")) {
            return ExpenseCategory.TRANSFER;
        }
        return ExpenseCategory.OTHER;
    }

    private boolean containsAny(String text, String... terms) {
        for (String term : terms) {
            if (text.contains(term)) {
                return true;
            }
        }
        return false;
    }
}

File: sms-expense-tracker/app/src/main/java/com/srijayant/expense/data/ExpenseParser.kt
package com.srijayant.expense.data

/**
 * Parses bank / UPI debit SMS into structured expenses.
 * Tuned for common Indian bank alert formats (HDFC, SBI, ICICI, Axis, Kotak, Paytm, PhonePe, GPay).
 */
object ExpenseParser {

    private val debitKeywords = listOf(
        "debited", "spent", "paid", "purchase", "withdrawn",
        "withdrawal", "sent", "transferred to", "payment of",
        "txn of", "transaction of", "dr ", " dr."
    )

    private val creditKeywords = listOf(
        "credited", "received", "deposited", "refund", "cashback", "cr "
    )

    private val amountPatterns = listOf(
        Regex("""(?i)(?:rs\.?|inr|₹)\s*([0-9]{1,3}(?:,[0-9]{2,3})*(?:\.[0-9]{1,2})?|[0-9]+(?:\.[0-9]{1,2})?)"""),
        Regex("""(?i)(?:amount|amt|of)\s*(?:rs\.?|inr|₹)?\s*([0-9]{1,3}(?:,[0-9]{2,3})*(?:\.[0-9]{1,2})?|[0-9]+(?:\.[0-9]{1,2})?)"""),
        Regex("""(?i)(?:debited|spent|paid|sent)\s*(?:by|for|with)?\s*(?:rs\.?|inr|₹)?\s*([0-9,]+\.?[0-9]*)""")
    )

    private val merchantPatterns = listOf(
        Regex("""(?i)(?:at|to|towards|for)\s+([A-Za-z0-9&.\-\s]{2,40}?)(?:\s+on\s|\s+via\s|\s+using\s|\s+upi|\s+ref|\s+avl|\.|,|$)"""),
        Regex("""(?i)(?:VPA|UPI ID)[:\s]+([A-Za-z0-9.\-_@]+)"""),
        Regex("""(?i)(?:Info|Merchant)[:\s]+([A-Za-z0-9&.\-\s]{2,40})""")
    )

    private val categoryRules: List<Pair<ExpenseCategory, List<String>>> = listOf(
        ExpenseCategory.FOOD to listOf(
            "swiggy", "zomato", "dominos", "mcdonald", "kfc", "starbucks",
            "cafe", "restaurant", "dining", "food", "blinkit", "zepto", "instamart"
        ),
        ExpenseCategory.SHOPPING to listOf(
            "amazon", "flipkart", "myntra", "ajio", "meesho", "nykaa",
            "reliance", "dmart", "bigbasket", "store", "mart"
        ),
        ExpenseCategory.TRAVEL to listOf(
            "uber", "ola", "rapido", "irctc", "makemytrip", "goibibo",
            "indigo", "airindia", "petrol", "fuel", "parking", "metro", "fastag"
        ),
        ExpenseCategory.BILLS to listOf(
            "electricity", "bses", "tata power", "airtel", "jio", "vi ",
            "vodafone", "broadband", "gas", "water", "recharge", "billdesk",
            "rent", "emi", "insurance", "premium"
        ),
        ExpenseCategory.ENTERTAINMENT to listOf(
            "netflix", "spotify", "prime", "hotstar", "disney", "youtube",
            "bookmyshow", "pvr", "inox", "sony liv"
        ),
        ExpenseCategory.HEALTH to listOf(
            "pharmacy", "apollo", "1mg", "pharmeasy", "hospital", "clinic",
            "diagnostic", "medlife", "netmeds"
        ),
        ExpenseCategory.ATM to listOf("atm", "cash wdl", "cash withdrawal", "withdrawn"),
        ExpenseCategory.TRANSFER to listOf("neft", "imps", "rtgs", "upi/", "transfer", "sent to")
    )

    data class ParsedSms(
        val amount: Double,
        val merchant: String,
        val category: ExpenseCategory,
        val isDebit: Boolean
    )

    fun parse(body: String, sender: String = ""): ParsedSms? {
        val text = body.replace('\n', ' ').trim()
        if (text.isBlank()) return null

        val lower = text.lowercase()
        val looksLikeDebit = debitKeywords.any { lower.contains(it) }
        val looksLikeCredit = creditKeywords.any { lower.contains(it) }

        // Prefer debits; skip pure credits (income/refunds) for expense tracking
        if (!looksLikeDebit && !looksLikeCredit) {
            // Still try if amount + bank-like sender
            if (!isLikelyBankSender(sender) || extractAmount(text) == null) return null
        }
        if (looksLikeCredit && !looksLikeDebit) return null

        val amount = extractAmount(text) ?: return null
        if (amount <= 0.0 || amount > 10_000_000) return null

        val merchant = extractMerchant(text) ?: guessMerchantFromSender(sender)
        val category = categorize(merchant, text)

        return ParsedSms(
            amount = amount,
            merchant = merchant,
            category = category,
            isDebit = true
        )
    }

    private fun extractAmount(text: String): Double? {
        for (pattern in amountPatterns) {
            val match = pattern.find(text) ?: continue
            val raw = match.groupValues.getOrNull(1)?.replace(",", "") ?: continue
            val value = raw.toDoubleOrNull() ?: continue
            if (value > 0) return value
        }
        return null
    }

    private fun extractMerchant(text: String): String? {
        for (pattern in merchantPatterns) {
            val match = pattern.find(text) ?: continue
            val raw = match.groupValues.getOrNull(1)?.trim().orEmpty()
            val cleaned = raw
                .replace(Regex("""\s+"""), " ")
                .trim(' ', '.', ',', '-', ':')
            if (cleaned.length in 2..40 && !cleaned.equals("your", ignoreCase = true)) {
                return cleaned.replaceFirstChar { it.titlecase() }
            }
        }
        return null
    }

    private fun guessMerchantFromSender(sender: String): String {
        val s = sender.uppercase()
        return when {
            "HDFC" in s -> "HDFC Bank"
            "SBI" in s || "SBIIN" in s -> "SBI"
            "ICICI" in s -> "ICICI Bank"
            "AXIS" in s -> "Axis Bank"
            "KOTAK" in s -> "Kotak Bank"
            "PAYTM" in s -> "Paytm"
            "PHONPE" in s || "PHONEPE" in s -> "PhonePe"
            "GPAY" in s || "GOOGLE" in s -> "Google Pay"
            "BOB" in s || "BARODA" in s -> "Bank of Baroda"
            "PNB" in s -> "PNB"
            else -> if (sender.isNotBlank()) sender else "Unknown"
        }
    }

    private fun isLikelyBankSender(sender: String): Boolean {
        val s = sender.uppercase()
        return listOf(
            "HDFC", "SBI", "ICICI", "AXIS", "KOTAK", "PAYTM", "PHONPE",
            "PHONEPE", "GPAY", "BOB", "PNB", "IDFC", "YESBK", "INDUS",
            "BANK", "UPI", "AMEX", "CITI"
        ).any { it in s }
    }

    fun categorize(merchant: String, rawMessage: String): ExpenseCategory {
        val haystack = "$merchant $rawMessage".lowercase()
        for ((category, keywords) in categoryRules) {
            if (keywords.any { haystack.contains(it) }) return category
        }
        return ExpenseCategory.OTHER
    }
}
